pymc/distributions/distribution.py

```diff
@@ -244,6 +275,21 @@ def fn(*args, **kwargs):
     return fn
 
 
+class _class_or_instancemethod(classmethod):
+    """Allow a method to be called both as a classmethod and an instancemethod,
+    giving priority to the instance method.
+
+    This is used to allow extracting information from the signature of a SymbolicRandomVariable
+    which may be provided either as a class attribute or as an instance attribute.
+
+    Adapted from https://stackoverflow.com/a/28238047
+    """
+
+    def __get__(self, instance, type_):
+        descr_get = super().__get__ if instance is None else self.__func__.__get__
+        return descr_get(instance, type_)
+
+
```

Comment:
Some bit of blackmagic to allow dynamic properties from the signature, to work both when signatures are class attributes or only instance attributes. Let me know if you see a better/ less magical solution.

Reply:
Why is it needed though?

Reply:
Sometimes we have the signature fixed in the class (DM), sometimes only known at runtime (Mixture). This allows us to parse ndims_params, ndim_supp, default_outputs from the signature regardless of when it's defined.

Reply:
Would it be better to have a class_signature (always defined) and a instance_signature (default None), and drop back to the class_signature if instance_signature is None?

Reply:
Do I really need this? Not so much. Alternatives are:
1. Hard code ndim_supp/ndims_params/default_output when they can be known in advance. It's just redundant because all info exists in the signature.
2. Don't provide any of these until the Op is defined. Drawback? I can't use these lines as straightfoward:

We have to create dummy Ops with empty size just for this to work. A lot of user dists fail with empty size
https://github.com/pymc-devs/pymc/blob/5f95fc292adebbeb31b6de825b7b407042e382fd/pymc/distributions/distribution.py#L486-L489

I can't write classmethods of the new SymbolicRandomVariables like this:
```python
        if rv_size_is_none(size):
            size = implicit_size_from_params(b, kappa, mu, ndims_params=cls.ndims_params)
```
Because cls.ndims_params doesn't exist until we create the `Op` below. I could obviously parse ndims_params on the spot from the signature.

It's just a small convenience.

Reply:
@jessegrabowski My last commit reverts the magic solution for the cleanest alternative I can think of. Which one do you prefer?

Reply:
I removed the non-magic commit, for reference it's here: https://github.com/pymc-devs/pymc/commit/20e6e3ba23b27020ce99a603394d8b67b8fef64a